backend/api/src/api/v1/gql/error.rs

```diff
@@ -0,0 +1,213 @@
+    fields: Vec<String>,
+    span: tracing::Span,
+    source: Option<anyhow::Error>,
+    location: &'static Location<'static>,
+}
+
+impl GqlErrorInterface {
+    fn with_source(self, source: Option<anyhow::Error>) -> Self {
+        Self { source, ..self }
+    }
+
+    fn with_location(self, location: &'static Location<'static>) -> Self {
+        Self { location, ..self }
+    }
+
+    pub fn with_field(self, fields: Vec<&str>) -> Self {
+        let fields = fields.into_iter().map(|f| f.to_string()).collect();
+        Self { fields, ..self }
+    }
+
+    pub fn with_message(self, message: String) -> Self {
+        Self {
+            message: Some(message),
+            ..self
+        }
+    }
+
+    fn display(&self) -> String {
+        match &self.message {
+            Some(msg) => format!("{}: {}", self.kind, msg),
+            None => format!("{}", self.kind),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq)]
+
+pub enum GqlError {
+    /// An internal server error occurred.
+    InternalServerError,
+    /// The input was invalid.
+    InvalidInput,
+    /// The session is no longer valid.
+    InvalidSession,
+    /// Not Implemented
+    NotImplemented,
+    /// Unauthorized
+    Unauthorized,
+}
+
+impl Display for GqlError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            GqlError::InternalServerError => write!(f, "InternalServerError"),
+            GqlError::InvalidInput => write!(f, "InvalidInput"),
+            GqlError::InvalidSession => write!(f, "InvalidSession"),
+            GqlError::NotImplemented => write!(f, "NotImplemented"),
+            GqlError::Unauthorized => write!(f, "Unauthorized"),
+        }
+    }
+}
+
+impl GqlError {
+    #[track_caller]
+    pub fn with_message(self, message: &str) -> GqlErrorInterface {
+        GqlErrorInterface {
+            kind: self,
+            message: Some(message.to_string()),
+            span: tracing::Span::current(),
+            source: None,
+            fields: Vec::new(),
+            location: Location::caller(),
+        }
+    }
+}
+
+impl ErrorExtensions for GqlErrorInterface {
+    fn extend(&self) -> async_graphql::Error {
+        let err = async_graphql::Error::new(self.display()).extend_with(|_, e| {
+            e.set("kind", self.kind.to_string());
+            if let Some(message) = &self.message {
+                e.set("reason", message.as_str());
+            }
+
+            e.set("fields", self.fields.as_slice());
+        });
+
+        match self.kind {
+            GqlError::InternalServerError => {
+                self.span.in_scope(|| {
+                    tracing::error!(error = ?self.source, location = self.location.to_string(), "gql error: {}", self.display());
+                });
+            }
+            _ => {
+                self.span.in_scope(|| {
+                    tracing::debug!(error = ?self.source, location = self.location.to_string(), "gql error: {}", self.display());
```

Comment:
Yes, but is is not implied since these are errors along our GQL API? It should be known that they are not standard GQL errors since this library does not implement the GQL spec. We use a different library to do that.